### ISSUE TITLE
DB schema dumper uncompatibility in Rails 7.0

### ISSUE THREAD

**Issue by mpg-tomohiko-mimura** [rails/rails#43909]
### Steps to reproduce

```sh
git@github.com:mito5525/precision_sample_app.git
cd precision_sample_app
bundle install

bin/rails db:drop db:create db:migrate
bin/rails db:schema:dump
git diff
# There is no difference.

bin/rails db:schema:load
bin/rails db:schema:dump
git diff
```

### Expected behavior

No diffs in schema.rb

### Actual behavior

The following differences will occur

```diff
diff --git a/db/schema.rb b/db/schema.rb
index 298ff3e..6d02e6b 100644
--- a/db/schema.rb
+++ b/db/schema.rb
@@ -13,7 +13,7 @@
 ActiveRecord::Schema.define(version: 2021_12_17_090228) do

   create_table "tasks", charset: "utf8mb4", force: :cascade do |t|
-    t.datetime "deleted_at"
+    t.datetime "deleted_at", precision: 6
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
   end
```

I think there was a lack of compatibility considerations in https://github.com/rails/rails/pull/42297.

### System configuration

**Rails version**: v7.0.0

**Ruby version**: v3.0.3

**Comment by morgoth** [comment]
Same here.

1. We dump schema from production database backup (so this is our source of truth)
2. The schema contains columns as `timestamp without time zone` as this is how it was in previous versions and is in our production db
3. From this schema.rb we bootstrap local database, which creates columns that are `timestamp(6) without time zone` so not really what is defined in the schema
4. On the `db:schema:dump` the `precision: 6` is added

@robertomiranda do you have a clue maybe what could cause it in https://github.com/rails/rails/pull/42297 ?

**Comment by bfad** [comment]
We are also seeing an issue here. Our `schema.rb` doesn't specify a precision for MySQL `datetimes`, and none in set in our databases. However, if we drop and reload our databases in Rails 7, a precision of 6 is added to the MySQL fields, but the schema.rb file remains unchanged.

**Comment by bfad** [comment]
Looks like #44171 may help fix this issue.

**Comment by morgoth** [comment]
The merged PR was about mysql. We're using postgresql, so I think this still will be an issue

**Comment by rafaelfranca** [comment]
Before loading the schema with a new version of Rails you should dump it. I'm going to fix this requirement, but right now that is the correct way to solve the issue.

**Comment by rafaelfranca** [comment]
Long term fix is here https://github.com/rails/rails/pull/44286

**Comment by justisb** [comment]
> Before loading the schema with a new version of Rails you should dump it. I'm going to fix this requirement, but right now that is the correct way to solve the issue.

I don't think that would solve this issue- because the old default `precision: nil` is not written in the dump. So there is no way to represent the Rails 6.1 non-timestamps datetime default precision in schema.rb, as the lack of precision keyword causes `db:schema:load` to always use `precision: 6` in Rails 7.0.

I believe your pending PR that allows specifying the compatibility version of the Schema class will fix it- that solution looks great. But I've found no simple temporary fix, and have currently resorted to monkey patching `ActiveRecord::Migration` with the `Compatibility::V6_1` definitions.

**Comment by rafaelfranca** [comment]
> I don't think that would solve this issue- because the old default `precision: nil` is not written in the dump.

This is already fixed in 7-0-stable.

**Comment by justisb** [comment]
> > I don't think that would solve this issue- because the old default `precision: nil` is not written in the dump.
> 
> This is already fixed in 7-0-stable.

I see the patch https://github.com/rails/rails/commit/eca89fd0241faa176cb7f9ca5dc09705e849a0be from PR #44171 on 7-0-stable, but that will only affect MySQL and not Postgres, right? Whereas your PR should effectively solve for any DB adapter after setting the schema.rb version to 6.1. Unless you're referring to a different fix.

**Comment by rafaelfranca** [comment]
Postgres should not have that problem. Does it?

**Comment by justisb** [comment]
Here's a minimal reproduction with PG, forked from the MySQL example in the issue description:
https://github.com/justisb/precision_sample_app

```
bin/rails db:drop db:create db:migrate
bin/rails db:schema:dump
git diff
# No diff
```

```
bin/rails db:drop db:create db:schema:load
bin/rails db:schema:dump
git diff
diff --git a/db/schema.rb b/db/schema.rb
index 7664977..5e6ac9c 100644
--- a/db/schema.rb
+++ b/db/schema.rb
@@ -16,7 +16,7 @@ ActiveRecord::Schema.define(version: 2021_12_17_093703) do
   enable_extension "plpgsql"
 
   create_table "tasks", force: :cascade do |t|
-    t.datetime "deleted_at"
+    t.datetime "deleted_at", precision: 6
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
   end
```

**Comment by rafaelfranca** [comment]
Wait, that makes no sense at all. You don't even loaded the schema, so the precision in the database should not be 6. Are those really the steps you took?

**Comment by morgoth** [comment]
@rafaelfranca The steps here https://github.com/rails/rails/issues/43909#issuecomment-1015455733 are reproducing it for PostgreSQL

**Comment by rafaelfranca** [comment]
So steps to fix you problem are:

```
# Rails 6.1
bin/rails db:schema:load

# Rails 7.0
bin/rails db:schema:dump
# Commit the generated file
```

Or dump the production database using Rails 7.0.

**Comment by rafaelfranca** [comment]
Ok. I think I got the problem now. https://github.com/rails/rails/pull/44356 should fix it. Can someone try?

**Comment by rafaelfranca** [comment]
Ok, even #44356 would not fix it. https://github.com/rails/rails/pull/44358 should.

**Comment by thisismydesign** [comment]
@rafaelfranca I have an issue with `postgresql`. When we migrated to Rails 7 `precision: 6` got added to timestamps in our schema. It worked fine. After updating to `7.0.2` when I run `db:migrate` this is replaced with `precision: nil`. After this, one of our spec comparing dates is failing because of what looks like a precision related error:
```
Diff:
       @@ -1 +1 @@
       -Mon, 21 Feb 2022 14:10:49.640199900 CET +01:00
       +Mon, 21 Feb 2022 14:10:49.640199000 CET +01:00
```

More importantly, I think this means an actual schema change that would not be reflected on production, only locally.
